### PR TITLE
Improve pre-commit to generate Airflow diagrams as a code

### PR DESCRIPTION
**Pull request by potiuk**
Since we are getting more diagrams generated in Airflow using the "diagram as a code" approach, this PR improves the pre-commit to be more suitable to support generation of more of the images coming from different sources, placed in different directories and generated independently, so that the whole process is more distributed and easy for whoever creates diagrams to add their own diagram.

The changes implemented in this PR:

* the code to generate the diagrams is now next to the diagram they generate. It has the same name as the diagram, but it has the .py extension. This way it is immediately visible where is the source of each diagram (right next to each diagram)

* each of the .py diagram Python files is runnable on its own. This way you can easily regenerate the diagrams by running corresponding Python file or even automate it by running "save" action and generate the diagrams automatically by running the Python code every time the file is saved. That makes a very nice workflow on iterating on each diagram, independently from each othere

* the pre-commit script is given a set of folders which should be scanned and it finds and run the diagrams on pre-commmit. It also creates and verifies the md5sum hash of the source Python file separately for each diagram and only runs diagram generation when the source file changed vs. last time the hash was saved and committed. The hash sum is stored next to the image and sources with .md5sum extension

Also updated documentation in the CONTRIBUTING.rst explaining how to generate the diagrams and what is the mechanism of that generation.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
This one should set us on a path where we should be able to convert pretty much all our architecture-like-diagrams to diagrams as a code approach easily.

It also allows to iterate on the diagrams very easily and explains how to integrate diagram "live preview" making it possible to instantly regenerate each diagram separately using "save actions".

File: scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py [hunk 1]
```diff
@@ -34,201 +30,28 @@
...
-        hash_file.write_text(str(my_file_hash) + "\n")
-    else:
-        console.print("[bright_blue]No changes to generation script. Not regenerating the images.")
+    for directory in DOCS_IMAGES_DIR, FAB_PROVIDER_DOCS_IMAGES_DIR:
+        diagram_files = directory.glob("diagram_*.py")
+        for source_file in diagram_files:
+            checksum = _get_file_hash(source_file)
+            hash_file = source_file.with_suffix(".md5sum")
+            if not hash_file.exists() or not hash_file.read_text().strip() == str(checksum).strip():
+                console.print(f"[bright_blue]Changes in {source_file}. Regenerating the image.")
```
Comment (josh-fell) [inline scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py]: Should we add license generation here for the md5sum file or ignore licenses for these types of files?

**Review by josh-fell — COMMENTED** [review]
This split is nice and more scalable indeed.

File: scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py (hunk 1 above)
Comment (potiuk) [reply scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py]: Generated files do not need licences :)

Comment (potiuk) [reply scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py]: So yeah. Adding to ignerefile should be ok

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -981,6 +981,58 @@ Documentation for ``apache-airflow`` package and other packages that are closely
...
+static checks below) but they can also be generated manually by running the corresponding Python code.
+
+To run the code you need to install the dependencies in the virtualenv you use to run it:
+* ``pip install diagrams rich``
+
+You need to have graphviz installed in your system (``brew install graphviz`` on macOS for example).
+
+The source code of the diagrams are next to the generated diagram, the difference is that the source
+code has ``.py`` extension and the generated diagram has ``.png`` extension. The pre-commit hook
+we have ``generate-airflow-diagrams`` will look for ``diagram_*.py`` files in the directories where it
```
Comment (vincbeck) [inline CONTRIBUTING.rst]:
```suggestion
``generate-airflow-diagrams`` will look for ``diagram_*.py`` files in the directories where it
```

File: docs/apache-airflow-providers-fab/img/diagram_fab_auth_manager_airflow_architecture.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (vincbeck) [inline docs/apache-airflow-providers-fab/img/diagram_fab_auth_manager_airflow_architecture.py:1]: You mixed the two diagrams. This one goes to Airflow doc and the other goes here in the fab provider (look at the images, they are not the same)

File: scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py [hunk 4]
```diff
@@ -34,201 +30,28 @@
...
-        generate_basic_airflow_diagram()
-        generate_dag_processor_airflow_diagram()
-        generate_auth_manager_airflow_diagram()
-        os.chdir(FAB_PROVIDER_DOCS_IMAGES_DIR)
-        generate_fab_auth_manager_airflow_diagram()
-        os.chdir(DOCS_IMAGES_DIR)
-        hash_file.write_text(str(my_file_hash) + "\n")
-    else:
-        console.print("[bright_blue]No changes to generation script. Not regenerating the images.")
+    for directory in DOCS_IMAGES_DIR, FAB_PROVIDER_DOCS_IMAGES_DIR:
```
Comment (vincbeck) [inline scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py]: Would not it be better to set `pass_filenames` to `True` in `.pre-commit-config.yaml` and loop over these files (ignoring md5 files). This way, no need to hardcode the different directories

File: docs/apache-airflow-providers-fab/img/diagram_fab_auth_manager_airflow_architecture.py (hunk 3 above)
Comment (potiuk) [reply docs/apache-airflow-providers-fab/img/diagram_fab_auth_manager_airflow_architecture.py:1]: 🤦

File: scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py (hunk 4 above)
Comment (potiuk) [reply scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py]: Yeah might be better and simpler, the problem with such approach is that we would have to agree that `diagram_*.py` is the convention that will be universally not used for something else. But I think it's pretty safae to assume this when we limit ourselves to `docs` folder.

**Comment by vincbeck** [comment]
Feel free to merge without my approval (I am now away until next year)

File: CONTRIBUTING.rst (hunk 2 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: Resoleved.

File: scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py (hunk 4 above)
Comment (potiuk) [reply scripts/ci/pre_commit/pre_commit_generate_airflow_diagrams.py]: Done

File: docs/apache-airflow-providers-fab/img/diagram_fab_auth_manager_airflow_architecture.py (hunk 3 above)
Comment (potiuk) [reply docs/apache-airflow-providers-fab/img/diagram_fab_auth_manager_airflow_architecture.py:1]: Solved.

**Comment by potiuk** [comment]
> Feel free to merge without my approval (I am now away until next year)

Have good holidays :)

**Review by josh-fell — APPROVED** [review]

**Review by vincbeck — APPROVED** [review]

**Pull request merged**